### PR TITLE
Enhancements to Quick Add Personal Vocab Feature

### PR DESCRIPTION
**Pull request by Tanvez**
[ATGU-3145](https://jira.huit.harvard.edu/browse/ATGU-3145) and [ATGU-3161](https://jira.huit.harvard.edu/browse/ATGU-3161)
- added select options for related nodes in modal
- added test for django and vue
- added preferred language
- make node selection accessible via arrow keys

Screen Recording 2021-07-29 at 2 07 25 PM

Update:
- refactor lattice_nodes api to look at lemma_nodes when building the lattice_nodes list relating to a headword
- refactor regex to include trailing numbers in results
- updated vue select options for suggested gloss to strip out numbers from lattice_node labels
- updated django test in relation to updated api

### CODE REVIEW

**Review by dodget — APPROVED** [review]
Great work Vesna! I'd like to answer a couple of questions before merging, mainly around adding the preferred language to the user profile.

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        ('lemmatized_text', '0014_auto_20201012_2328'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='lemmatizedtext',
+            name='cloned_for',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='groups.group'),
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (dodget) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: This rogue migration could be the result of django drift. Probably nothing that you did.

File: hedera/api.py [hunk 2]
```diff
@@ -312,17 +319,20 @@ def get_data(self):
...
+
+    def get_data(self):
+        headword = self.request.GET.get("headword")
+        filtered_headword_iterable = filter(str.isalnum, headword)
+        filtered_headword_string = "".join(filtered_headword_iterable)
+        qs = LatticeNode.objects.filter(label__iregex=r"\y" + re.escape(filtered_headword_string) + r"\y")
+        lattice_node_list = []
+        for node in qs:
+            lattice_node_list.append(node.to_dict())
+        return lattice_node_list
```
[the remark this answers is not in the document]
Comment (Tanvez) [reply hedera/api.py:338]: oh I didn't know that! Thats a cool feature! Will update the code with those changes.

File: hedera/migrations/0004_profile_lang.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        ('hedera', '0003_profile_show_node_ids'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='profile',
+            name='lang',
+            field=models.CharField(max_length=3, null=True),
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (Tanvez) [reply hedera/migrations/0004_profile_lang.py:18]: Yea, definitely! Will hold off on any merging until 👍 by Bill.

File: hedera/tests/test_api.py [hunk 4]
```diff
@@ -20,8 +25,45 @@ def test_post_personal_vocabulary_quick_add_api(self):
...
+class MeAPITest(APITestCase):
+    def setUp(self):
+        self.created_user = User.objects.create_user(username=f"test_user{randrange(100)}", email=f"test_user{randrange(100)}@test.com", password="password")
+
+    def test_post_me_profile_with_lang_field(self):
+
+        self.client.force_login(user=self.created_user)
+        response = self.client.post("/api/v1/me/", json.dumps({"lang": "lat"}), content_type="application/json")
+        data = response.json()["data"]
+        self.assertEqual(data["lang"], "lat")
```
[the remark this answers is not in the document]
Comment (Tanvez) [reply hedera/tests/test_api.py:69]: ahh will update the test with failing tests!

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tanvez) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: It was generated when I ran the migrations for the Profile model, not sure if it was a previous change that was never migrated or django drift 🤔

File: hedera/tests/test_api.py [hunk 5]
```diff
@@ -54,16 +94,27 @@ def test_get_related_lattice_nodes(self):
...
+
+    def test_successful_post_me_profile_with_lang_field(self):
         response = self.client.post("/api/v1/me/", json.dumps({"lang": "lat"}), content_type="application/json")
         data = response.json()["data"]
         self.assertEqual(data["lang"], "lat")
+
+    def test_failing_post_me_profile_with_lang_field(self):
+        # if language is not supported
+        response = self.client.post("/api/v1/me/", json.dumps({"lang": "abc"}), content_type="application/json")
+        self.assertEqual(response.status_code, 400)
```
Comment (dodget) [inline hedera/tests/test_api.py:120]: 👍

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  } from '../../constants';
+  import FamiliarityRating from '../../modules/FamiliarityRating.vue';
+
+  export default {
+    components: { FamiliarityRating },
+    // on creation of the dom element fetch the list of langauages/ids the user has in their personal vocab list
+    async created() {
+      await this.$store.dispatch(FETCH_PERSONAL_VOCAB_LANG_LIST);
+      await this.$store.dispatch(FETCH_ME);
+      window.addEventListener('keyup', this.onKey);
```
Comment (Tanvez) [inline static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:127]: todo: remove dead code

File: hedera/api.py [hunk 7]
```diff
@@ -310,19 +320,29 @@ def get_data(self):
...
-#         headword = self.request.GET.get("headword")
-#         filtered_headword_iterable = filter(str.isalnum, headword)
-#         filtered_headword_string = "".join(filtered_headword_iterable)
-#         qs = LatticeNode.objects.filter(label__iregex=r"\y"+ re.escape(filtered_headword_string) + r"\y")
-#         data = serializers.serialize('json', qs)
-#         json_data = json.loads(data)
-#         return json_data
+        checked_data = self.check_data(data)
+        if checked_data is not True:
+            return HttpResponseBadRequest("Missing required fields")
```
Comment (arthurian) [inline hedera/api.py:334]: Should we return these 400 responses as JSON, that way the client knows the errors will be returned as JSON as well? Maybe something like this (just a suggestion):

```python
class JsonResponseBadRequest(JsonResponse):
    status_code = 400
```

And then you could write:

```python
return JsonResponseBadRequest(data={"error": "Missing required fields"})
```

File: hedera/api.py [hunk 8]
```diff
@@ -310,19 +320,29 @@ def get_data(self):
...
+        return JsonResponse({"data": {"created": True, "data": new_entry.data()}})
+
+
+class LatticeNodesAPI(APIView):
+
+    def get_data(self):
+        headword = self.request.GET.get("headword")
+        filtered_headword_iterable = filter(str.isalnum, headword)
+        filtered_headword_string = "".join(filtered_headword_iterable)
+        qs = LatticeNode.objects.filter(label__iregex=r"\y" + re.escape(filtered_headword_string) + r"\y")
```
Comment (arthurian) [inline hedera/api.py:347]: Just curious, but what is `\y` in the regex?

File: hedera/migrations/0004_profile_lang.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arthurian) [reply hedera/migrations/0004_profile_lang.py:18]: Agreed on getting input from @bilbe.

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arthurian) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: Yeah, I suspect this is drift. And as we discussed offline, we should pin it at django 3.2 now so that doesn't happen again (at least not without our knowing).

File: static/src/js/app/Dashboard.vue [hunk 9]
```diff
@@ -1,28 +1,13 @@
...
-    <Modal
-      ref="Modal"
-      title="Add Word to Personal Vocabulary List"
-      description="Modal to quick add vocab"
-      ><QuickAddVocabForm ref="QuickAddVocabForm"
-    /></Modal>
-    <button type="button" class="btn btn-primary" @click="openModal">
-      Quick Add Vocab
-    </button>
+    UNDER CONSTRUCTION
```
Comment (arthurian) [inline static/src/js/app/Dashboard.vue:3]: I suggest we remove this, since we can assume it's under construction.

File: static/src/js/app/components/quick-add-button/QuickAddButton.vue [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+</template>
+<script>
+  import Modal from '../Modal.vue';
+  import QuickAddVocabForm from './QuickAddVocabForm.vue';
+  import { FETCH_ME } from '../../constants';
+
+  export default {
+    components: { Modal, QuickAddVocabForm },
+    async created() {
+      await this.$store.dispatch(FETCH_ME);
```
Comment (arthurian) [inline static/src/js/app/components/quick-add-button/QuickAddButton.vue:22]: I'm wondering if it would make sense to have the `FETCH_ME` action dispatched from the `Dashboard.vue` component instead of here. My assumption is that the "me" data only needs to be read once, when the dashboard is loaded. I could imagine other sub-components on the dashboard needing that data, and we wouldn't want every sub-component dispatching the same action. So if the dashboard managed that state by dispatching the action, then sub-components wouldn't need to manage it, they could just read the state directly from the store. I could be thinking about this wrong, so curious what you think.

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+          headword: this.headword,
+        });
+        // sets first latticenode as the default in select options
+        if (this.latticeNode) {
+          this.latticeNodeId = this.latticeNode[0].pk;
+        }
+      },
+      onSelect(event) {
+        // note: '+' changes the string to a number
+        const node = this.latticeNode.find((ele) => +ele.pk === +event.target.value);
```
Comment (arthurian) [inline static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:221]: 👍  for the note about the `+`. That's a neat trick, although you could make the case that `parseInt()` is clearer and reflects the expected format of the number.

**Review by arthurian — APPROVED** [review]
Nice work, @Tanvez! It's great to see how the QuickAdd components are evolving, along with the backend and frontend unit tests. Kudos for making those better.

Overall this looks like it's in good shape to merge, other than a few questions around the preferred language that @dodget mentioned. I added a few questions/suggestions, but nothing that should block you from moving forward.

File: hedera/api.py (hunk 7 above)
Comment (Tanvez) [reply hedera/api.py:334]: Yea, we definitely should! Will update the code to reflect this.

File: hedera/api.py (hunk 8 above)
Comment (Tanvez) [reply hedera/api.py:347]: Think it might have something to do with how Postgres uses word boundaries. I found an old stackoverflow post here [PostgreSQL Regex Word Boundaries](https://stackoverflow.com/questions/3825676/postgresql-regex-word-boundaries/29798772). We might have to tweak this to accommodate the changes Ivy was talking about.

File: static/src/js/app/Dashboard.vue (hunk 9 above)
Comment (Tanvez) [reply static/src/js/app/Dashboard.vue:3]: Will do!

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue (hunk 11 above)
Comment (Tanvez) [reply static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:221]: ahh that would be more descriptive! will update with the `parseInt()`

File: hedera/api.py (hunk 8 above)
Comment (Tanvez) [reply hedera/api.py:347]: Should we hold off on refactoring the regex or refactor the regex to account for numerals in the headword?

File: static/src/js/app/components/quick-add-button/QuickAddButton.vue (hunk 10 above)
Comment (Tanvez) [reply static/src/js/app/components/quick-add-button/QuickAddButton.vue]: I think due to how hedera loads in different vue components periodically some of the vuex state might not persist between page navigation. Since we are planning to add this button site wide, might need to be updated in [ATGU-3160](https://jira.huit.harvard.edu/browse/ATGU-3160)

Comment (arthurian) [reply static/src/js/app/components/quick-add-button/QuickAddButton.vue]: Yeah, it may require further refactoring when we get to that story.

File: hedera/api.py (hunk 8 above)
Comment (Tanvez) [reply hedera/api.py:347]: @jtauber We had a question about the following when looking up headwords in relation to the personal vocabulary list:
- if we wanted to search for a headword should we look at `lattices_latticenodes`  table
- how are the headword stored based on multiple meanings between the same headword

File: hedera/api.py [hunk 12]
```diff
@@ -393,8 +393,34 @@ def get_data(self):
...
+        lattice_node_list = []
+        for node in lattice_nodes:
+            gloss = node["gloss"]
+            if gloss != "from morpheus" and gloss != "morpheus ambiguity":
+                lattice_node_list.append(node)
+            elif len(node["children"]):
+                # checks child nodes with matching label and appends them to the result
+                for child_node in node["children"]:
+                    valid_lemma = bool(re.search(rf"\b{headword}[0-9]*\b", child_node["label"]))
+                    if child_node["gloss"] != "from morpheus" and child_node["gloss"] != "morpheus ambiguity" and valid_lemma is not False:
```
Comment (Tanvez) [inline hedera/api.py:411]: todo: leave off check for valid_lemma
